bsip-0069.md

```diff
@@ -0,0 +1,165 @@
+
+# Specifications
+
+#### account_authorities_match_predicate:
+
+```c++
+/**
+ * This predicate takes one or more authority or special_authority objects and
+ * asserts that each supplied authority matches the corresponding account
+ * authority.  For authorities NOT supplied, no checks are performed.
+ */
+struct account_authorities_match_predicate
+{
+   account_id_type account_id;
+
+   optional<authority>  owner;
+   optional<authority>  active;
+   optional<special_authority>  owner_special;
+   optional<special_authority>  owner_active;
+
+   /**
+    *  Validate that predicate is well-formed.  Verify that at least one optional
+    *  parameter is supplied and that they are valid (special_)authority objects.
+    */
+   bool validate() const;
+};
+
+```
+
+#### asset_issuer_is_account_predicate:
+
+```c++
+/**
+ * Asserts that asset_id->issuer matches issuer_id.
+ */
+struct asset_issuer_is_account_predicate
+{
+   asset_id_type   asset_id;
+   account_id_type issuer_id;
+
+   bool validate() const;
+};
+
+```
+
+#### asset_precision_predicate:
+
+```c++
+/**
+ * Assert that asset_id->precision matches precision.
+ */
+struct asset_precision_predicate
+{
+   asset_id_type   asset_id;
+   uint8_t         precision;
+
+   bool validate() const;
+};
+
+```
+
+#### head_block_time_ge_time_predicate:
+
+```c++
+/**
+ * Asserts that the block time of the block which this transaction would
+ * build upon (i.e. the parent of the block this tx would be included in)
+ * meets or exceeds a specified time.  This allows to create transactions
+ * which only become valid “in the future”. (Verifies that
+ * database::head_block_time() >= min_time.)
+ */
+struct head_block_time_ge_time_predicate
+{
+   fc::time_point_sec min_time;
+
+   bool validate() const;
+};
+
+```
+
+# Discussion
+
+### Impacted committee parameters
+
+An affected committee parameter is `max_predicate_opcode`, which limits the number of defined predicates supported by the assert operation.  If new predicate types are added to the core protocol, this parameter will need to be increased in order to “activate” the new predicates on the network.  Thus, the committee should be ready to increase this parameter if this BSIP finds support and is implemented.  Using a committee parameter to activate protocol additions after block-producer adoption is an alternative to setting a hard-fork date for activation.
```

Comment:
Nice find.

IMO we should do away with that parameter. We still will have to add hardfork protection because old nodes can't deserialize the new predicates, and the existing check does not prevent new assertions from being included in proposals.

The parameter currently has value "1" but we have three predicates defined (see above). The implementation of the third (block_id) is incomplete, which is probably the reason why that parameter exists. Question is - do we want to remove the broken block_id predicate, or fix it? @abitmore ?

Reply:
> Please add the removal of the check to the "Specification" section. What do you think @abitmore ?

Would like @ryanRfox and perhaps others' takes on this.  We had some discussion on this during a previous core call.  I feel a little apprehensive "stealing" a committee parameter away from them.  But... if it's reason for being is no longer relevant,... then I guess so.  Is it appropriate to "bundle" removal of this parameter into this BSIP?  Or should it be it's own?

Reply:
The predicates need to be approved by BSIP approval vote (Community) and protocol upgrade code (Block Producers), there fore I do not see the need for the Committee to also "activate" them. Removing this ability is intentional, but I feel reflects the process we currently use for all other protocol upgrades. Perhaps we should seek some comments from current Committee members such as @bitcrab @xeroc @roelandp @OpenLedgerApp @clockworkgr

Reply:
I agree with @ryanRfox 
The committee has nothing to do with blockchain features perse.

Reply:
Text has been updated to propose removal of `max_predicate_opcode`.